### DISCUSSION TITLE
Observer for Variants

### DISCUSSION THREAD

**Discussion by jefflew** [zencart/zencart#6401] [Ideas]
Can you please also add a parameter to the zen_get_products_stock and zen_check_stock functions called $attributes?  This way we can use the observer for managing variant stock.

**Comment by lat9** [comment]
Neither of those functions "know" anything about the requested product's currently-selected attributes.
```php
function zen_get_products_stock($products_id) ...

function zen_check_stock($products_id, $products_quantity) ...
```

**Comment by drbyte** [comment]
> use the observer for managing variant stock

Is this in relation to a public addon/plugin? 

I was considering some other potential notifiers, and seeing your logic around variant-handling would be helpful to inform the decision-making.

**Comment by jefflew** [comment]
There's already an observer in both functions.  It just needs an array of
attributes added to it and then we can manage variant stock without editing
the core function.

**Reply by lat9** [reply]
... and where might that array of attributes be sourced?  As I indicated above, the function has absolutely no knowledge of what attribute-variant is associated with the active request; it knows only the `products_id`.

**Reply by lat9** [reply]
My bad, the `$products_id` input to each of those functions is either the associated product's `id` or a `prid` (an `id` suffixed with the md5 hash of its attributes).  If the observer of the notification sees that the value's the latter, the observer can look-up in the `$_SESSION['cart']`'s `contents` array to find the array of associated attributes.

**Comment by jefflew** [comment]
Yes, we could use the existing observer to connect to the session.  However
we may be wanting to use it prior to the product being in the cart, such as
when displaying inventory on a product page or checking stock before
adding. We can then use that function and pass it the attributes.  Still
requires core edits in some areas, but no longer needing to edit the
function itself will be a big help and easier to maintain.

**Reply by lat9** [reply]
Can't help with the product-page inventory, but for prior-to-adding to the cart, you can check in the same notifications for
```php
if (isset($_GET['action']) && $_GET['action'] === 'add_product' && isset($_POST['products_id']) && $_POST['products_id'] == $p1 && isset($_POST['id']) && is_array($_POST['id'])) {
```
... and use the `$_POST['id']` array as your source for the to-be-added variant.

**Reply by mc12345678** [reply]
While haven't reviewed for use when product not in and not being added to the cart, but [Stock by Attributes](https://github.com/mc12345678/Stock_By_Attributes_Combined) as of Zen Cart 1.5.6 collects the attribute information using the existing notifier and without core code modification.
